### PR TITLE
adapt to changes in core

### PR DESCRIPTION
**Pull request by kba**
* adapted to changes in core
* word segmenter (within lines only atm)

### CODE REVIEW

File: ocrd-tool.json [hunk 1]
```diff
@@ -18,7 +18,20 @@
       "tags": ["Text recognition and optimization"],
       "description": "Recognize text in lines with tesseract",
```
Comment (wrznr) [inline ocrd-tool.json]: Is it lines only?

File: ocrd-tool.json [hunk 2]
```diff
@@ -18,7 +18,20 @@
...
       "executable": "ocrd-tesserocr-recognize",
-      "step": "recognition/text-recognition"
+      "step": "recognition/text-recognition",
+      "parameter": {
+        "textequiv_level": {
```
Comment (wrznr) [inline ocrd-tool.json]: I am not sure here: Shouldn't we propagate the `textequiv` contents through all the levels?

File: ocrd_tesserocr/recognize.py [hunk 3]
```diff
@@ -1,41 +1,49 @@
...
-                        # xTODO use binarized / gray
-                        image = self.workspace.resolve_image_as_pil(image_url, line.coords)
-                        tessapi.SetImage(image)
-                        line.textequiv = tessapi.GetUTF8Text()
+                pcgts = from_file(self.workspace.download_file(input_file))
+                # TODO use binarized / gray
+                pil_image = self.workspace.resolve_image_as_pil(pcgts.get_Page().imageFilename)
+                tessapi.SetImage(pil_image)
+                # TODO slow
+                #  tessapi.SetPageSegMode(PSM.SINGLE_LINE)
```
Comment (wrznr) [inline ocrd_tesserocr/recognize.py]: We definitely have to check what's wrong here.

File: ocrd_tesserocr/segment_line.py [hunk 4]
```diff
@@ -12,19 +18,27 @@ def process(self):
...
-                page = OcrdPage.from_file(self.workspace.download_file(input_file))
-                image_url = page.imageFileName
-                for region in page.list_textregions():
-                    log.debug("Detecting lines in %s with tesseract", region)
-                    image = self.workspace.resolve_image_as_pil(image_url, region.coords)
+                pcgts = from_file(self.workspace.download_file(input_file))
+                image_url = pcgts.get_Page().imageFilename
+                for region in pcgts.get_Page().get_TextRegion():
+                    log.debug("Detecting lines in %s with tesseract", region.id)
+                    image = self.workspace.resolve_image_as_pil(image_url, polygon_from_points(region.get_Coords().points))
```
Comment (wrznr) [inline ocrd_tesserocr/segment_line.py]: Wasn't this function renamed?

File: ocrd-tool.json (hunk 1 above)
Comment (kba) [reply ocrd-tool.json]: At the moment it is. Walk the regions, walk the lines within the regions, recognize text in the bounding box of the line. To make that configurable was the idea of the textequiv_level parameter.

File: ocrd-tool.json (hunk 2 above)
Comment (kba) [reply ocrd-tool.json]: Yes, `TextEquiv` should be propagated where possible. But you might want to run the recognition on sub-regions, on different levels. If the input PAGE is already zoned on line level, you can (as you implemented) set the PSM to `SINGLE_LINE`. Or you might have an OCR result on the word/glyph level as input and want to recognize only the words to compare results.

Maybe `recognition-level` is less confusing?

File: ocrd_tesserocr/segment_line.py (hunk 4 above)
Comment (kba) [reply ocrd_tesserocr/segment_line.py]: The functions have the syntax `X_from_Y`, where X/Y can be
  * `points` a string encoding a polygon: `"0,0 100,0 100,100, 0,100"`
  * `polygon` an array of x-y-tuples of a polygon: `[[0,0], [100,0], [100,100], [0,100]]`
  * `xywh` a dict with keys for x, y, width and height: `{'x': 0, 'y': 0, 'w': 100, 'h': 100}`

`points` is for PAGE

`polygon` is what opencv2 expects

`xywh` is what tesserocr expects/produces.

**Pull request merged**